parlai/mutators/flip_labels.py

```diff
@@ -0,0 +1,140 @@
+work with binary classification only.
+"""
+
+
+@register_mutator("flip_classification_label")
+class FlipClassificationLabelMutator(MessageMutator):
+    @classmethod
+    def add_cmdline_args(
+        cls, parser: ParlaiParser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        super().add_cmdline_args(parser, partial_opt)
+        agent = parser.add_argument_group('FlipLabelMutator args')
+        agent.add_argument(
+            '--noise-level',
+            type=float,
+            help='The probability of label flipping',
+        )
+        return parser
+
+    def __init__(self, opt: Opt):
+        super().__init__(opt)
+        if opt['noise_level'] > 1 or opt['noise_level'] < 0:
+            raise Exception('Noise_level should be within 0 - 1.')
+        if opt['noise_level'] == 0:
+            logging.info('No noise, teacher should be the same as before.')
+        if opt['noise_level'] == 1:
+            logging.info('Labels are exact the opposite.')
+        self.noise_level = opt['noise_level']
+        logging.info(f'Flipping with noise level {self.noise_level}')
+        self.is_training = DatatypeHelper.is_training(opt['datatype'])
+        self.flip_cnt = 0
+        return
+
+    def message_mutation(self, message: Message) -> Message:
+        message = copy.deepcopy(message)
+        if 'labels' in message:
+            labels = message['labels']
+            text = message['text']
+            new_labels = []
+            flip = hashlib.md5(text.encode('utf8')).digest()[0] < self.noise_level * 256
+            if len(labels) != 1:
+                raise ValueError(
+                    f'{type(self).__name__} can only be used with one label!'
+                )
+            else:
+                # deterministic flipping for each example
+                if flip:
+                    self.flip_cnt += 1
+                    if labels[0] == '__ok__':
+                        new_labels = ['__notok__']
+                    elif labels[0] == '__notok__':
+                        new_labels = ['__ok__']
+                    else:
+                        raise ValueError("labels must be '__ok__' and '__notok__'")
+                    assert len(new_labels) == 1
+            if flip:
+                message.force_set('labels', new_labels)
+        return message
+
+    def __del__(self):
+        logging.info(f'Flipped {self.flip_cnt}')
+
+
+@register_mutator("flip_classification_label_train_only")
+class FlipClassificationLabelTrainOnlyMutator(FlipClassificationLabelMutator):
+    def __init__(self, opt: Opt):
+        super().__init__(opt)
+        if not self.is_training:
+            logging.info('No flipping in eval / test mode.')
+        return
+
+    def message_mutation(self, message: Message) -> Message:
+        if not self.is_training:
+            return message
+        else:
+            return super().message_mutation(message)
+
+
+@register_mutator("flip_label_valid_only")
```

Comment:
maybe rename it as "flip_classification_label_valid_only" to be consistent with L85